casper/abstract_view.py

```diff
@@ -41,65 +41,57 @@ def _next_display_height(self):
-            return
-
         for message in showed_messages:
             if message.hash in self.pending_messages or message.hash in self.justified_messages:
                 continue
 
-            missing_message_hashes = self.get_missing_messages_in_justification(message)
+            missing_message_hashes = self.missing_messages_in_justification(message)
             if not any(missing_message_hashes):
-                newly_justified_messages = {message}
-                newly_justified_messages.update(self.resolve_waiting_messages(message))
-                self.add_to_justified_messages(newly_justified_messages)
+                self.add_to_justified_messages(message)
+                self.resolve_waiting_messages(message)
+                return
 
-                for message_hash in self.pending_messages:
-                    assert message_hash in self.missing_message_dependencies
-            else:
-                for missing_message_hash in missing_message_hashes:
-                    if missing_message_hash not in self.dependents_of_message:
-                        self.dependents_of_message[missing_message_hash] = []
-
-                    self.dependents_of_message[missing_message_hash].append(message.hash)
-                    self.missing_message_dependencies[message.hash] = missing_message_hashes
-                    self.pending_messages[message.hash] = message
+            self.pending_messages[message.hash] = message
+            self._track_missing_messages(message, missing_message_hashes)
 
     def resolve_waiting_messages(self, message):
         """Given a new message, resolve all messages that are waiting for it to be justified"""
         if message.hash not in self.dependents_of_message:
-            return set()
+            return
 
-        newly_justified_messages = set()
-        for message_hash in self.dependents_of_message[message.hash]:
+        for dependent_hash in self.dependents_of_message[message.hash]:
             # sanity check!
-            assert message.hash in self.missing_message_dependencies[message_hash]
+            assert message.hash in self.missing_message_dependencies[dependent_hash]
 
-            self.missing_message_dependencies[message_hash].remove(message.hash)
+            self.missing_message_dependencies[dependent_hash].remove(message.hash)
 
-            if not any(self.missing_message_dependencies[message_hash]):
-                new_message = self.pending_messages[message_hash]
-                newly_justified_messages.add(new_message)
-                newly_justified_messages.update(self.resolve_waiting_messages(new_message))
-                del self.missing_message_dependencies[message_hash]
-                del self.pending_messages[message_hash]
+            if not any(self.missing_message_dependencies[dependent_hash]):
+                new_message = self.pending_messages[dependent_hash]
+                self.add_to_justified_messages(new_message)
```

Comment:
Thoughts about keeping this function as side-effects free as possible? I originally had this here as well, but it seemed like it might be harder to reason about it. But I guess they are fairly similar in any case...

Reply:
eh, this shit is going to have side effects... even if we aren't adding messages to the justified dict, we are still deleting things from missing_message_dependences and pending_messages. Seems wrong to pull something off of `pending_messages` and not put it onto the `justified_messages` in a semi-atomic action.

Reply:
i found both functions a lot more readable in the current format